specs/deposit_gas_metering.md

```diff
@@ -0,0 +1,153 @@
+
+- [Deposits on L1](#deposits-on-l1)
+  - [Guaranteed Gas](#guaranteed-gas)
+  - [Gas Stipend](#gas-stipend)
+  - [Additional Gas](#additional-gas)
+  - [Limiting Guaranteed Gas](#limiting-guaranteed-gas)
+  - [Rationale for Guaranteed vs Additional Gas](#rationale-for-guaranteed-vs-additional-gas)
+  - [Rationale for burning L1 Gas](#rationale-for-burning-l1-gas)
+- [Deposits on L2](#deposits-on-l2)
+  - [Guaranteed Gas On L2](#guaranteed-gas-on-l2)
+  - [Additional Gas On L2](#additional-gas-on-l2)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Deposits on L1
+
+L1 Deposits are augmented with the following fields:
+
+- `guaranteedGas`
+- `gasPrice`
+- `additionalGas`
+
+`guaranteedGas` and `additionalGas` must fit into a 64 bit unsigned integer.  
+`gasPrice` must fit into a 256 bit unsigned integer.
+
+### Guaranteed Gas
+
+The **guaranteed gas** is composed of a gas stipend and any additional guaranteed
+gas the user would like to buy on L1.
+
+If the user requests more `guaranteedGas` than the `gasStipend`, that gas will
+be bought with L1 Eth via a gas burn or by buying it directly.
+It is not refundable if the transaction uses less gas than the gas limit.
+
+### Gas Stipend
+
+Can also provide a stipend to the guaranteed gas of every transaction.
+The stipend is bought with gas that was spent executing the deposit logic;
+however, the stipend still needs to be included in the sum of bought
+guaranteed gas.
+
+TODO: How much / do we actually need this?
+
+If a gas stipend is provided, we only buy the amount of guaranteed gas
+in excess of the gas stipend.
+
+### Additional Gas
+
+Users can request additional gas for their L2 transaction. This gas is
+refundable is less than the gas limit is used. However, if there is not
+enough available gas on L2 or the gas price not high enough, the additional
+gas will not be provided on L2.
+
+Additional gas is bought with the following:
+
+- `gasPrice`: L2 gas price to buy.
+- `additionalGas`: Amount of additional gas requested
+
+The additional gas will be bought with the `from` account's balance. We do
+not require that the user mint enough L2 eth to cover the additional gas fee,
+but do suggest that tooling checks that the user has enough balance for it.
+
+### Limiting Guaranteed Gas
+
+The deposit feed contract must limit the total amount of guaranteed gas in a
+single L1 block. This is to limit the amount of gas that is used on L2.
+
+This is done with a hard limit on the amount of guaranteed gas that can be
+bought. To reduce PGAs if the deposit mechanism is congested, we also implement
+an EIP-1559 style fee market with the following pseudo code:
+
+```text
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 2
+
+curr_basefee: u128, curr_num: u64, curr_bought_gas: u64 = load_and_unpack_storage()
+GUARANTEED_GAS_LIMIT: u64, SANITY_GAS_LIMIT: u64 = load_and_unpack_storage2()
+gas_target = GUARANTEED_GAS_LIMIT // ELASTICITY_MULTIPLIER
```

Comment:
What are those `load_and_unpack_storage()` for? (Feels like related to ABI?) Does it really need to be part of the spec (and if so, can it be explained better?)

Reply:
What is the `SANITY_GAS_LIMIT`?

Reply:
Must also explain what `curr_bought_gas` is (i.e. requested guaranteed gas on top of stipend).